modules/tinymce/src/plugins/link/test/ts/browser/SelectedTextLinkTest.ts

```diff
@@ -32,14 +56,104 @@ UnitTest.asynctest('browser.tinymce.plugins.link.SelectedTextTest', (success, fa
+            'a[href="http://something"]': 1,
+            'strong:contains(word)': 1,
+            'em:contains(other)': 1,
+            'p': 1
+          })
+        )
+      ]),
+      Log.stepsAsStep('TINY-5205', 'Link: complex selections with existing link should replace the text when changing text', [
+        tinyApis.sSetContent('<p><a href="http://oldlink/"><strong>word</strong><em>other</em></p>'),
+        tinyApis.sSetSelection([ 0 ], 0, [ 0 ], 1),
+        sOpenDialog(),
+        FocusTools.sSetActiveValue(doc, 'http://something'),
+        Keyboard.sKeydown(doc, Keys.tab(), { }),
+        FocusTools.sSetActiveValue(doc, 'new text'),
+        TestLinkUi.sClickSave,
+        Waiter.sTryUntil(
+          'Wait until link is inserted',
+          tinyApis.sAssertContentPresence({
+            'a[href="http://something"]': 1,
+            'a:contains(new text)': 1,
+            'strong': 0,
+            'em': 0,
+            'p': 1
+          })
+        )
+      ]),
+      Log.stepsAsStep('TINY-5205', 'Link: collapsed selection in complex structure should preserve the text when changing URL', [
+        tinyApis.sSetContent('<p><a href="http://oldlink/"><strong>word</strong><em>other</em></p>'),
+        tinyApis.sSetSelection([ 0, 0, 0, 0 ], 2, [ 0, 0, 0, 0 ], 2),
+        sOpenDialog(),
+        FocusTools.sSetActiveValue(doc, 'http://something'),
+        TestLinkUi.sClickSave,
+        Waiter.sTryUntil(
+          'Wait until link is inserted',
+          tinyApis.sAssertContentPresence({
+            'a[href="http://something"]': 1,
+            'strong:contains(word)': 1,
+            'em:contains(other)': 1,
+            'p': 1
+          })
+        )
+      ]),
+      Log.stepsAsStep('TINY-5205', 'Link: Selection with link inside should replace link', [
+        tinyApis.sSetContent('<p>a <a href="http://www.google.com/">b</a> c</p>'),
+        tinyApis.sSetSelection([ 0, 0 ], 0, [ 0, 2 ], 2),
+        sOpenDialog(),
+        FocusTools.sSetActiveValue(doc, 'http://something'),
+        TestLinkUi.sClickSave,
+        Waiter.sTryUntil(
+          'Wait until link is updated',
+          tinyApis.sAssertContentPresence({
+            'a[href="http://something"]': 1,
+            'a': 1,
+            'p': 1
+          })
+        )
+      ]),
+      Log.stepsAsStep('TINY-5205', 'Link: Selection across partial link should split and replace link', [
+        tinyApis.sSetContent('<p><a href="http://www.google.com/">a b</a> c</p>'),
+        tinyApis.sSetSelection([ 0, 0, 0 ], 2, [ 0, 1 ], 2),
+        sOpenDialog(),
+        FocusTools.sSetActiveValue(doc, 'http://something'),
+        TestLinkUi.sClickSave,
+        Waiter.sTryUntil(
+          'Wait until link is updated',
+          tinyApis.sAssertContentPresence({
+            'a[href="http://www.google.com/"]': 1,
+            'a[href="http://something"]': 1,
+            'a': 2,
+            'p': 1
+          })
+        )
+      ]),
+      TestLinkUi.sClearHistory
```

Comment:
Does this need to happen between tests, or just at the start / end?

Reply:
Just the end technically (left start in for consistency), as none of these tests rely on the autocompleter.